File: BBSPICE/BBSPICE/Mathematics/Matrix/LU.c
//
//  LU.c
//  BBSPICE
//
//  Created by Oleksandr Bolbat on 05.03.2026.
//

#ifndef LU_H
#include "LU.h"


int LU_Decomposition(double *LU, const size_t size, size_t *permutation) {
    for (size_t row = 0; row < size; ++row) permutation[row] = row;
    
    for (size_t row = 0; row < size; ++row) {
        size_t p = row;
        double max = fabs(LU[row * size + row]);
        for (size_t column = row + 1; column < size; ++column) {
            const double cell = fabs(LU[column * size + row]);
            if (cell > max) {
                max = cell;
                p = column;
            }
        }
        
        if (max < EPS) return DIVISION_SINGULAR;
        
        if (p != row) {
            for (size_t column = 0; column < size; ++column) {
                double temp = LU[row * size + column];
                LU[row * size + column] = LU[p * size + column];
                LU[p * size + column] = temp;
            }
            size_t temp = permutation[row];
            permutation[row] = permutation[p];
            permutation[p] = temp;
        }
        
        for (size_t column = row + 1; column < size; ++column) {
            LU[column * size + row] /= LU[row * size + row];
            for (size_t i = row + 1; i < size; ++i) {
                LU[column * size + i] -= LU[column * size + row] * LU[row * size + i];
            }
        }
    }
    
    return DIVISION_SUCCES;
}

int LU_Solve(const double *LU, const size_t size, const size_t *permutation,
             const double *I, double *V, double *buffer) {
    
    for (size_t row = 0; row < size; ++row) buffer[row] = I[permutation[row]];
    
    for (size_t row = 0; row < size; ++row) {
        double sum = buffer[row];
        for (size_t column = 0; column < row; ++column) sum -= LU[row * size + column] * buffer[column];
        buffer[row] = sum;
    }
    
    for (ptrdiff_t ii = (ptrdiff_t) size - 1; ii >= 0; --ii) {
        size_t row = (size_t) ii;
        double sum = buffer[row];
        
        for (size_t column = row + 1; column < size; ++column) sum -= LU[row * size + column] * V[column];
        
        double uii = LU[row * size + row];
        if (fabs(uii) < EPS) return DIVISION_SINGULAR;
        
        V[row] = sum / uii;
    }
    
    return DIVISION_SUCCES;
}


#endif

File: BBSPICE/BBSPICE/Mathematics/Matrix/LU.swift
//
//  LU.swift
//  BBSPICE
//
//  Created by Oleksandr Bolbat on 06.03.2026.
//

import Foundation


#if canImport(LU)
import LU
#else
@_silgen_name("LU_Decomposition")
func LU_Decomposition(_ LU: UnsafeMutablePointer<Double>?, _ size: Int, _ permutation: UnsafeMutablePointer<UInt>?) -> Int32
@_silgen_name("LU_Solve")
func LU_Solve(_ LU: UnsafeMutablePointer<Double>?, _ size: Int, _ permutation: UnsafeMutablePointer<UInt>?,
              _ I: UnsafeMutablePointer<Double>?, _ V: UnsafeMutablePointer<Double>?, _ buffer: UnsafeMutablePointer<Double>?) -> Int32
#endif



func LU_Division(_ G: [Double], _ I: inout [Double], _ size: Int) -> [Double] {
    var LU = G
    var permutation = Array<UInt>(repeating: 0, count: size)
    var buffer = Array<Double>(repeating: 0.0, count: size)
    var V = Array<Double>(repeating: 0.0, count: size)
    
    let err1 = LU.withUnsafeMutableBufferPointer { LU_Ptr -> Int32 in
        permutation.withUnsafeMutableBufferPointer { PermutationPtr -> Int32 in
            return Int32(LU_Decomposition(LU_Ptr.baseAddress, size, PermutationPtr.baseAddress))
        }
    }
    
    precondition(err1 == 0, "LU_Decomposition Failed")
    
    let err2 = LU.withUnsafeMutableBufferPointer { LU_Ptr -> Int32 in
        permutation.withUnsafeMutableBufferPointer { PermutationPtr -> Int32 in
            V.withUnsafeMutableBufferPointer { V_Ptr -> Int32 in
                I.withUnsafeMutableBufferPointer { I_Ptr -> Int32 in
                    buffer.withUnsafeMutableBufferPointer { BufferPtr -> Int32 in
                        return Int32(LU_Solve(LU_Ptr.baseAddress, size, PermutationPtr.baseAddress,
                                              I_Ptr.baseAddress, V_Ptr.baseAddress, BufferPtr.baseAddress))
                    }
                }
            }
        }
    }
    
    precondition(err2 == 0, "LU_Solve Failed")
    
    return V
}
